migrate_cpp/cpp_refactoring/rewriter_test.cpp

```diff
@@ -0,0 +1,133 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "migrate_cpp/cpp_refactoring/rewriter.h"
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "clang/AST/ASTConsumer.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendAction.h"
+#include "clang/Tooling/Tooling.h"
+
+namespace Carbon::Testing {
+namespace {
+
+// Represents C++ source code with at most one region enclosed in $[[...]]$ as
+// an annotated range.
+class Annotations {
+ public:
+  Annotations(std::string_view annotated_source) {
+    size_t index = annotated_source.find("$[[");
+    if (index == std::string_view::npos) {
+      source_code = std::string(annotated_source);
+      return;
+    }
+    start = index;
+    end = annotated_source.find("]]$", index);
+    source_code = (llvm::Twine(annotated_source.substr(0, start)) +
+                   annotated_source.substr(start + 3, end - start - 3) +
+                   annotated_source.substr(end + 3))
+                      .str();
+    // Update `end` so that it is relative to the unannotated source (which
+    // means three characters earlier due to the `$[[` being removed.
+    end -= 3;
+  }
+
+  // Returns a view into the unannotated source.
+  std::string_view source() const { return source_code; }
+
+  // Returns the offsets in the file representing the annotated range if they
+  // exist and `{0, std::numeric_limits<size_t>::max()}` otherwise.
+  std::pair<size_t, size_t> range() const { return std::pair(start, end); }
+
+ private:
+  std::string source_code;
+  size_t start = 0;
+  size_t end = std::numeric_limits<size_t>::max();
+};
+
+// Rewrites the `cpp_code`, return the Carbon equivalent. If the text has no
+// source range annotated with $[[...]]$, the entire translation unit will be
+// migrated and output. Otherwise, only the migrated output corresponding to the
+// annotated range will be be output. No more than one range may be annoated at
+// all.
+//
+// This annotation mechanism is useful in that it allows us to specifically test
+// the migration associated with specific nodes even when they require some
+// additional context that we do not wish to be covered by the test.
+auto RewriteText(std::string_view cpp_code) -> std::string {
+  std::string result;
+
+  Annotations annotated_cpp_code(cpp_code);
+
+  bool success = clang::tooling::runToolOnCodeWithArgs(
+      std::make_unique<MigrationAction>(result, annotated_cpp_code.range()),
+      annotated_cpp_code.source(), {}, "test.cc", "clang-tool",
+      std::make_shared<clang::PCHContainerOperations>(),
+      clang::tooling::FileContentMappings());
+
```

Comment:
```suggestion
  EXPECT_TRUE(success) << "Tool failed.";
```

I think something like this would be helpful for explicitly flagging the source of errors; also, it's probably worth thinking about how the tool should return errors so that error messages can be tested, too.

Regardless, returning the empty string may add to confusion for people new to the codebase.

Reply:
I generally prefer not to have expectations inside functions (only inside TEST). How would you feel about returning an ErrorOr? Do we have GTest matchers for these available yet? If not, how do you feel about handling this in a follow-up?

Reply:
I'd be fine with that, I don't think there are matchers at present.